### PR TITLE
build(icon): added icon generator script

### PR DESCRIPTION
**Pull request by DonatelloTHM**
Description:
This PR introduces a `generate-icons.ts` script to be used for icon generation.
- Added a new script (`generateSvgIcons`) to handle the generation of icon `svg` sprites and icon list/types
- Added an Icon component that utilizes the generated `svg` sprites.

Execute the script with `pnpm run build:icons`

Video:

https://github.com/femmecubator/equity-ui/assets/61029296/df3903ef-b1c6-4326-820f-0c7f66bcd97b

### CODE REVIEW

File: src/icons/scripts/generate-icons.ts [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+import { parse } from 'node-html-parser';
+import * as p from '@clack/prompts';
+import { optimize } from 'svgo';
+import chalk from 'chalk';
+import { Project, SyntaxKind, SourceFile } from 'ts-morph';
+import * as prettier from 'prettier';
+import { iconList } from '../icon-constant';
+import { configuration } from './svgo-config';
+
+const renderLogo = () => {
```
Comment (vuonga1103) [inline src/icons/scripts/generate-icons.ts]: can we use function declaration here for consistency?

File: src/icons/scripts/generate-icons.ts [hunk 2]
```diff
@@ -0,0 +1,465 @@
...
+      return;
+    },
+  });
+
+  if (p.isCancel(response)) onCancel();
+
+  return response.toString();
+}
+
+renderLogo();
```
Comment (vuonga1103) [inline src/icons/scripts/generate-icons.ts]: let's move this up top to after imports

File: src/icons/scripts/generate-icons.ts [hunk 3]
```diff
@@ -0,0 +1,465 @@
...
+  const initializer = exportConst
+    .getInitializerIfKindOrThrow(SyntaxKind.AsExpression)
+    .getExpressionIfKindOrThrow(SyntaxKind.ArrayLiteralExpression);
+  const elementIndex = iconList.findIndex((elem) => elem === iconName);
+  initializer.removeElement(elementIndex);
+
+  await formatAndSaveSourceFile(sourceFile, project);
+}
+
+generateSvgIcons();
```
Comment (vuonga1103) [inline src/icons/scripts/generate-icons.ts]: let's move this up top to after imports

File: src/icons/scripts/generate-icons.ts [hunk 4]
```diff
@@ -0,0 +1,465 @@
...
+}
+
+async function handleSpecificIconGeneration(
+  inputDir: string,
+  outputDir: string
+) {
+  const files = glob
+    .sync('**/*.svg', { cwd: inputDir })
+    .sort((a, b) => a.localeCompare(b));
+  const filteredFiles = files.filter((file) => file.includes('24'));
```
Comment (vuonga1103) [inline src/icons/scripts/generate-icons.ts]: we should discuss with Design/Krizia re: having the Figma icons exported in the default `[name].svg` format so we don't have to deal with the weirdness of `Size=24, Name=Wellness.svg` naming convention

File: src/icons/scripts/generate-icons.ts [hunk 5]
```diff
@@ -0,0 +1,465 @@
...
+  if (filteredFiles.length === 0) {
+    p.note(`No SVG Icons found in ${inputDir}`);
+    onCancel();
+  }
+
+  const newIcons = filteredFiles.filter(
```
Comment (vuonga1103) [inline src/icons/scripts/generate-icons.ts]: nit

```suggestion
    const newIcons = filteredFiles.filter(
    (file) => {
      const iconName = convertString(path.basename(file, '.svg'));
      return !existingIcons.has(iconName)}
  );
```

**Review by vuonga1103 — CHANGES_REQUESTED** [review]
this is really really awesome! i am learning a lot about script writing through this work, so thank you. ❤️

in addition to the below comments, as discussed we could also:
1. refactor script to work without presence of `icon-constant.ts` (i think this is a super nice-to-have if you ever want to open source the script)
2. refactor `generate-icons.ts` to improve cleanliness/reusability, i.e. breaking the large file down to multiple modules; this could be done later as well.

File: src/icons/scripts/generate-icons.ts (hunk 4 above)
Comment (DonatelloTHM) [reply src/icons/scripts/generate-icons.ts]: Sounds good.

File: src/icons/scripts/generate-icons.ts (hunk 3 above)
Comment (DonatelloTHM) [reply src/icons/scripts/generate-icons.ts]: Bet

File: src/icons/scripts/generate-icons.ts (hunk 2 above)
Comment (DonatelloTHM) [reply src/icons/scripts/generate-icons.ts]: Bet

File: src/icons/scripts/generate-icons.ts (hunk 1 above)
Comment (DonatelloTHM) [reply src/icons/scripts/generate-icons.ts]: Aight bet

**Review by vuonga1103 — APPROVED** [review]

File: src/icons/icon.tsx [hunk 6]
```diff
@@ -0,0 +1,27 @@
+import { type SVGProps } from 'react';
+import spriteHref from './sprite.svg';
+import type { IconName } from './icon-constant';
+
+const defaultIconSizes = {
+  small: 18,
+  large: 24,
+};
```
Comment (limoncello526) [inline src/icons/icon.tsx:8]: let's use as const here

```suggestion
const defaultIconSizes = {
  small: 18,
  large: 24,
} as const;
```

File: src/icons/icon.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+export function Icon({
+  name,
+  size = 'large',
+  ...props
+}: SVGProps<SVGSVGElement> & {
+  name: IconName;
+  size?: DefaultIconSizes | number;
+}) {
+  const actualSize = typeof size === 'string' ? defaultIconSizes[size] : size;
```
Comment (limoncello526) [inline src/icons/icon.tsx:20]: let's use a type predicate instead

```suggestion

const isDefaultIconSize = (size: unknown): size is DefaultIconSizes => size === 'small' || size === 'large';

export function Icon({
  name,
  size = 'large',
  ...props
}: SVGProps<SVGSVGElement> & {
  name: IconName;
  size?: DefaultIconSizes | number;
}) {
  const actualSize = isDefaultIconSize(size) ? defaultIconSizes[size] : size;
```

File: src/icons/scripts/svgo-config.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
+import { type Config } from 'svgo';
```
Comment (limoncello526) [inline src/icons/scripts/svgo-config.ts:1]:
```suggestion
import type { Config } from 'svgo';
```

File: src/icons/scripts/svgo-config.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    {
+      name: 'preset-default',
+      params: {
+        overrides: {
+          removeViewBox: false,
+        },
+      },
+    },
+  ],
+};
```
Comment (limoncello526) [inline src/icons/scripts/svgo-config.ts:53]: let's use satisfies here

```suggestion
export const configuration = {
  plugins: [
    'cleanupAttrs',
    'removeDoctype',
    'removeXMLProcInst',
    'removeComments',
    'removeMetadata',
    'removeTitle',
    'removeDesc',
    'removeUselessDefs',
    'removeEditorsNSData',
    'removeEmptyAttrs',
    'removeHiddenElems',
    'removeEmptyText',
    'removeEmptyContainers',
    'cleanupEnableBackground',
    'convertStyleToAttrs',
    'convertStyleToAttrs',
    'convertColors',
    'convertPathData',
    'convertTransform',
    'removeUnknownsAndDefaults',
    'removeNonInheritableGroupAttrs',
    'removeUnusedNS',
    'cleanupIds',
    'cleanupNumericValues',
    'moveElemsAttrsToGroup',
    'moveGroupAttrsToElems',
    'collapseGroups',
    'removeRasterImages',
    'mergePaths',
    'convertShapeToPath',
    'sortAttrs',
    'removeDimensions',
    'removeXMLNS',
    {
      name: 'convertColors',
      params: {
        currentColor: true,
      },
    },
    {
      name: 'preset-default',
      params: {
        overrides: {
          removeViewBox: false,
        },
      },
    },
  ],
} satisfies Config;
```

**Review by limoncello526 — CHANGES_REQUESTED** [review]
minor changes

++ ~~Can we also add a storybook file~~ thx

File: .husky/pre-commit
Comment (limoncello526) [inline .husky/pre-commit:1]: @vuonga1103 : i think it maybe a good idea if we add the lint check in this workflow - https://github.com/femmecubator/equity-ui/blob/main/.github/workflows/pr-workflow.yaml

TODO

File: src/icons/scripts/svgo-config.ts (hunk 9 above)
Comment (DonatelloTHM) [reply src/icons/scripts/svgo-config.ts:53]: @cfernando415 The use of satisfies Config might be unnecessary in this context. We're not relying on type inference from configuration, we just need type type safety and Intellisense. 

P.S I'm not opposed to this change, I just think that reaching out for `satisfies` is mostly beneficial when we need both type safety and more specific inference. I'd like to hear your thoughts

File: src/icons/scripts/svgo-config.ts (hunk 8 above)
Comment (DonatelloTHM) [reply src/icons/scripts/svgo-config.ts:1]: I'm not opposed to this change, as both of those statements would be dropped on the JS output. I primarily use named type imports for clarity, as they immediately identify types and accommodate future non-type imports without updating the statement.
Ex. 
```ts
import { type Config, loadConfig } from 'svgo';
```

File: src/icons/icon.tsx (hunk 7 above)
Comment (DonatelloTHM) [reply src/icons/icon.tsx:20]: @cfernando415 Using a type predicate here might be counterproductive.

Consider this example:

```ts
<Icon size="somestring" name="arrow"/>
```

If a type predicate is used, `actualSize` will be set to `somestring`, which is invalid. However, with the existing method, `defaultIconSizes['somestring']` results in `undefined`, making `actualSize` also undefined (a valid entry).

This issue is already prevented by the type of size, unless there's a type casting. But the present method ensures both type-safety and runtime safety.

Please let me know what you think.

**Comment by DonatelloTHM** [comment]
@cfernando415 A story file will be added on a separate PR, when tokens are available, as we need color tokens for a complete story

File: src/icons/icon.tsx (hunk 7 above)
Comment (limoncello526) [reply src/icons/icon.tsx:20]: hmmm.. thanks for explaining! I understand the intention 😄 

resolving this thread - no change needed here

File: src/icons/scripts/svgo-config.ts (hunk 9 above)
Comment (limoncello526) [reply src/icons/scripts/svgo-config.ts:53]: hmmm... that is a valid point in this context and no other properties are being used that would require type safety as you mentioned.

I am good with that reasoning - no change needed for this

resolving this thread

File: .husky/pre-commit
Comment (vuonga1103) [reply .husky/pre-commit:1]: yes, good idea

**Review by limoncello526 — APPROVED** [review]

**Pull request merged**